Update python_plus_gis.py

SGJJSR opened the pull request:
my solution for Exercise 4

SGJJSR committed:
Update python_plus_gis.py

nicolaslara commented:
I can comment on your pull request! You did do it right :)

nicolaslara commented on python_plus_gis.py:
```diff
@@ -117,3 +117,22 @@ def createPolyGeom(coords):
+
+lines = []
+def linestringing (a, b):
+    for p, pp in zip(orig_points, dest_points):        
```
give these better names! It's hard to follow the code when they're called `p` and `pp`

nicolaslara commented on python_plus_gis.py:
```diff
@@ -117,3 +117,22 @@ def createPolyGeom(coords):
+        i = l.length
+        j += i
+    avg_len = j/len(linestring_list)
+    return(avg_len)
```
this code (`avg_distance()`) is correct and works. However, can you think of a more concise way of doing this? Maybe using functions that work on lists? maybe list comprehensions/generator expressions?

nicolaslara replied:
yeah, the change you did is what I was thinking of. You could make it even more concise by using a generator inside sum instead of generating a list:

```sum(l.length for l in linestring_list)/len(linestring_list)```

This is also more efficient for reasons that may not be worth for you to spend too much time on. I'm going to summarize them here, but don't worry too much about it since it's a more advanced topic:

 * Using `[x for x in ...] is a "list comprehension". It generates a list. 
 * Lists take space in memory. This is ok, we want them in memory; that's why we use lists: to store things.
 * Using `(x for x in ...)` (notice the parenthesis instead of brackets) is a "generator expression". This means: a short form of writing a generator. 
 * Generator expressions iterate over the elements the same way that lists comprehensions do, but don't store all elements in a resulting list. Instead, they `yield` the elements one by one to be consumed by whoever is calling them.
 * This means that using the generator expression is the same in terms of time and space as writing a for loop and processing each element inside. In the example above: the `sum()` function is the consumer, and it will add each element one by and then discard it (the original elements are still in `linestring_list` which isn't modified).

nicolaslara replied:
yeah, the change you did is what I was thinking of. You could make it even more concise by using a generator inside sum instead of generating a list:

```sum(l.length for l in linestring_list)/len(linestring_list)```

This is also more efficient for reasons that may not be worth for you to spend too much time on. I'm going to summarize them here, but don't worry too much about it since it's a more advanced topic:

 * Using `[x for x in ...] is a "list comprehension". It generates a list. 
 * Lists take space in memory. This is ok, we want them in memory; that's why we use lists: to store things.
 * Using `(x for x in ...)` (notice the parenthesis instead of brackets) is a "generator expression". This means: a short form of writing a generator. 
 * Generator expressions iterate over the elements the same way that lists comprehensions do, but don't store all elements in a resulting list. Instead, they `yield` the elements one by one to be consumed by whoever is calling them.
 * This means that using the generator expression is the same in terms of time and space as writing a for loop and processing each element inside. In the example above: the `sum()` function is the consumer, and it will add each element one by and then discard it (the original elements are still in `linestring_list` which isn't modified).

nicolaslara replied:
If you want to make it even shorter you could do

```
def avg_distance(linestrings)
    return sum(l.length for l in linestrings)/len(linestrings)
```

Notice: 
 1. No unnecessary variables
 2. It is pretty easy to understand what the function is doing: it basically describes what is doing when you read it out loud

A caveat. These "rules" (like not creating unnecessary variables) are not absolute, they depend of each case, and even when looking at the same code, different programmers will disagree (very strongly) about what's "cleaner". 

As an example, you may want to create "unnecessary" variables to avoid making the lines too long, or to give parts of the calculations names so that other parts read more easily.

SGJJSR committed:
is this version more efficient / better?

Thanks! Great tasks, I am learning a lot

SGJJSR committed:
updated the first function

I think this version actually spits out the true distance...(I had a conceptual flaw in my thinking).
I tried to summarize everything in a one line list comprehension (below), maybe you could judge if I did it correctly? it seems odd, I need to assign the result of the function to a new variable in order to make it work ( I didn't need to do that with the first solution).

nicolaslara commented on python_plus_gis.py:
```diff
@@ -117,3 +117,48 @@ def createPolyGeom(coords):
+
+def linestringing (a, b):
+    lines = [(LineString([(p1.x , p1.y),(p2.x, p2.y)])) for p1,p2 in zip(a,b)]        
+    return(lines)
```
I like this solution. No need to create `lines` just to return it, though
